### PR TITLE
Add functionality to insert, remove and replace layers in the model

### PR DESCRIPTION
**Pull request by soumyadipghosh**
- Adds insert, remove and replace functions to modify layers in the model
- Adds corresponding unit tests

NOTE: There is an assumption that the layer being modified has just one parent and one child.

### CODE REVIEW

**Comment by timmoon10** [comment]
[Bamboo is running](https://lc.llnl.gov/bamboo/browse/LBANN-TIM378-1).

**Review by samadejacobs — APPROVED** [review]
LGTM

File: src/models/model.cpp [hunk 1]
```diff
@@ -971,6 +971,129 @@ void model::add_split_layers(std::unordered_set<std::string>& layer_names) {
...
+  // Other checks also have to be done - like cannot delete activation function between two layers, etc
+
+  // Parent and child of removable layer
+  auto& parent = const_cast<Layer&>(l.get_parent_layer(0)); // assuming only one parent
+  auto& child = const_cast<Layer&>(l.get_child_layer(0)); // assuming only one child
+
+  // Setup relationship between parent layer and child layer
+  child.replace_parent_layer(l.get_parent_layer_pointer(0), child.find_parent_layer_index(l));
+  parent.clear_child_layers();
+  parent.add_child_layer(l.get_child_layer_pointer(0));
```
Comment (timmoon10) [inline src/models/model.cpp:1046]: We should handle the case where `parent` has multiple children:
```suggestion
  parent.replace_child_layer(l.get_child_layer_pointer(0), parent.find_child_layer_index(l));
```

File: src/models/model.cpp [hunk 2]
```diff
@@ -971,6 +971,129 @@ void model::add_split_layers(std::unordered_set<std::string>& layer_names) {
...
+  auto& parent = const_cast<Layer&>(l.get_parent_layer(0)); // assuming only one parent
+  auto& child = const_cast<Layer&>(l.get_child_layer(0)); // assuming only one child
+  
+  // Setup relationship between the new layer and child of old layer (which becomes child of new layer)
+  new_layer->add_child_layer(l.get_child_layer_pointer(0));
+  child.replace_parent_layer(new_layer, child.find_parent_layer_index(l));
+
+  // Setup relationship between parent of old layer (which becomes parent of new layer) and new layer
+  parent.clear_child_layers();
+  parent.add_child_layer(new_layer);
```
Comment (timmoon10) [inline src/models/model.cpp:1087]: We should handle the case where `parent` has multiple children:
```suggestion
  parent.replace_child_layer(new_layer, parent.find_child_layer_index(l));
```

**Review by timmoon10 — CHANGES_REQUESTED** [review]
Mostly looks good. Just an edge case if a parent layer has multiple children.

**Review by benson31 — APPROVED** [review]
Once @timmoon10 is happy, I'm ok with the rest of this.

**Review by timmoon10 — APPROVED** [review]

**Pull request merged**